File: c++/groupAnagrams.cpp

// leetcode.com : 49. Group Anagrams

#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Solution
{
  public:
    vector<vector<string>> groupAnagrams(vector<string> &strs)
    {
        vector<vector<string>> result;

        unordered_map<string, vector<string>> m;

        for (auto s : strs)
        {
            string t = s;

            sort(t.begin(), t.end());

            m[t].push_back(s);
        }

        for (auto s : m)
        {
            result.push_back(s.second);
        }

        return result;
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/GroupAnagrams.java
package com.vkeonline.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

/**
 * @author csgear
 * leetcode: 49. Group Anagrams
 */

public class GroupAnagrams {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> map = new HashMap<>();

        for (String s : strs) {
            char[] arr = s.toLowerCase().toCharArray();
            Arrays.sort(arr);
            String current = String.valueOf(arr);
            if (map.containsKey(current)) {
                List<String> list = map.get(current);
                list.add(s);
            } else {
                List<String> list = new ArrayList<String>();
                list.add((s));
                map.put(current, list);
            }
        }

        List<List<String>> result = new ArrayList<>(map.size());

        result.addAll(map.values());

        return result;
    }
}

